layer2: Make nodes re-join memberlist cluster

rata opened the pull request:
This PR makes the nodes join the memberlist cluster when they left for some reason (network outage, etc.).

It is based on the ideas from the RFC PR #647 but storing the Speaker IPs to also be able to recover when the API server is down (before we relied on the API server to gives us the APIs, so if it was down we couldn't join members), based on @champtar suggestion on that PR.

It consists of two commits: the first one is cherry-picked from #595 and just moves the memberlist code from `speaker/main.go` to it's own package, so it is easier to reason and abstract. The second one just adds a go routine that will try to re-join the nodes whenever it is needed.

The semantic changes in the second commit aim to be simple to follow and reason about, without any major rework. More reworks can be done in the future, in another minor release, taking into accounts the ideas from #595. This is aimed at v0.9.4, therefore the change tries to be as simple and correct as possible, avoiding major semantic changes.

I've tested this locally with kind (with `inv dev-env`) and blocking traffic with iptables. It worked well and as expected on all situations (API server down and recovers, one other node down and recovers, etc.).

Let me know what you think. 

TODO:
- [ ] Add tests. Memberlist is quite untested currently.

I'll c&p the commit message here from the rejoin commit, as it is the core of this PR. To properly understand what I refer to here, please look just at this commit and not the whole diff of the PR:

Currently, nodes join the memberlist cluster only on startup and if they are removed (e.g. small network downtime and they recover) they never join again. This means the node is alive, but not taken into account as a possible speaker, making it pointless to run the MetalLB speaker.

This commit adds a go routine to do memberlist joins in a loop. This was missing in the commit that introduced memberlist (3c5fc63 "Use hashicorp/memberlist to speedup dead node detection") and fixes issue #589.

This routine tries to join nodes in two situations: (a) every 1 minute, (b) when the current k8s node object changes and the SetNode() callback is called.

The case (b) is used when the node loses connectivity for long enough for Kubernetes to mark it in a Not Ready state. When the node is Ready again (e.g network is restored for the node), the callback is invoked and the node tries to join the cluster.

The case (a), periodic checks, is needed for downtimes that are smaller than the time Kubernetes takes to mark a node as Not Ready (about 45 sec by default). For example, if a node has a small downtime (~15 seconds) memberlist will remove it from the cluster but as this is short enough for Kubernetes to not change the node state to Not Ready, the SetNode() callback won't be invoked (as the node didn't change). In these cases, the periodic join every 1 minute will make those node join the
memberlist cluster again.

The internal kubernetes watcher already has this logic to notify when the current node changes. In the Layer2 case, though, it was just a "return nil". Now, it queues the message to process a re-join.

This routine tries to join only the nodes that are not currently part of the cluster, as it was advised by Etienne Champetier that the memberlist Join() operation is not a no-op for nodes that are already members and can take a considerable amount of time (specially when only a small group of nodes need to re-join the cluster).

Furthermore, the IPs to join are stored in the SpeakerList struct to be able to recover nodes even if the API server is down. This means we don't need to query the API server and just use the IPs cached in the struct. This was also very good advise from Etienne.

There is another go routine introduced here to update the IPs to join stored in the struct. This go routine updates them by quering the API server, every 5 minutes only to not overload the API server (something very easy to do by mistake). This is in a separate go routine because, as I mention in the code, the current version of client-go we are using doesn't support a context or a timeout so that query can block. To avoid this blocking other operations, we use a separate go routine. When we update to new client-go versions, this will of course not be needed and we can just run with a timeout if we want to.

johananl commented on internal/speakerlist/speakerlist.go:
```diff
@@ -0,0 +1,301 @@
+// IPs of all the speaker pods. As this queries the API server, it waits at
+// least 5 minutes between queries, to avoid self-induced API server DoS (should
+// be trated with care).
+func (sl *SpeakerList) speakersIPsLoop() {
```
The function name doesn't match the doc comment.

As for the naming here, `speakersIPsLoop()` sounds like describing the implementation to me. Better to name functions based on their functionality as seen by the caller.

rata replied:
Let's have the conversation on one thread only and then adjust them all. Let's continue here: https://github.com/metallb/metallb/pull/662#discussion_r452454917

johananl commented on internal/speakerlist/speakerlist.go:
```diff
@@ -0,0 +1,301 @@
+// IPs of all the speaker pods. As this queries the API server, it waits at
+// least 5 minutes between queries, to avoid self-induced API server DoS (should
+// be trated with care).
+func (sl *SpeakerList) speakersIPsLoop() {
```
Can't we leverage built-in functionality in the k8s library to avoid looping and waiting? I feel like we are re-implementing things which are provided by client-go.

We should probably look at using an informer (inside the `k8s` package) like we already do for watching other k8s objects.

rata replied:
Well, that is what #595 does, uses a service, and the ideas we discussed offline.

I'm not sure there is another way rather than creating a service for the speakers (which is not clear is something we want, as it is only needed for layer 2 and now the user needs to create a service) and then monitoring it. I'm short of time, so if you have the time to check if there are other alternatives, it would be great :)

An alternative in the path just mentioned (monitor a service), is to create a service an monitor it from the speakerlist package. Basically, we instantiate a `internal/k8s.go` monitoring. But we either modify it to accept a selector (currently monitors all services) or do things like get notified when a service changed and if it's not the metalLB speakers, ignore. As I mentioned offline, I think these ideas have merrit, but I really doubt about doing those changes in a patch release.

Basically it means all metalLB users need to create a service, and we haven't thought the design through to know if this is something that will stay or not. I'm really hesistant about it.

We added memberlist without any kind of re-join mechanism and now we need to fix it. I think to be safe, I lean towards fixing it with the primitives we already have. Unless we find a clean and simple way without exposing more things to the user (like having a service) that we might want to delete afterwards.

What do you think?

liuyuan10 replied:
I'm not sure if the client version used by metallb have it. Can we use NewFilteredListWatchFromClient with a list option that has the labelSelector?
https://godoc.org/k8s.io/client-go/tools/cache#NewFilteredListWatchFromClient
https://godoc.org/k8s.io/apimachinery/pkg/apis/meta/v1#ListOptions

champtar replied:
If I remember correctly you don't have the watch permission

rata replied:
@liuyuan10 cool, thanks! @johananl can you check those when you have some time? :)

johananl replied:
> I'm not sure if the client version used by metallb have it. Can we use NewFilteredListWatchFromClient with a list option that has the labelSelector?
> https://godoc.org/k8s.io/client-go/tools/cache#NewFilteredListWatchFromClient
> https://godoc.org/k8s.io/apimachinery/pkg/apis/meta/v1#ListOptions

Great pointers @liuyuan10, thanks!

Looks like we can do the following:

```go
om := func(o *metav1.ListOptions) {
	o.LabelSelector = "foo=bar"
}
watcher := cache.NewFilteredListWatchFromClient(c.client.CoreV1().RESTClient(), "pods", v1.NamespaceAll, om)
```

I'm looking into it.

rata replied:
We discussed offline with @johananl and we agreed that adding a watch mechanism without a proper design (so it might change) doesn't seem nice for a patch release. It is too big of a change to be safe, and without a design it seems really unsafe for a patch release.

For 0.9.4 we plan to just poll very gently, as this patch is doing.

johananl commented on internal/speakerlist/speakerlist.go:
```diff
@@ -0,0 +1,301 @@
+	return iplist, nil
+}
+
+func (sl *SpeakerList) mlJoinLoop() {
```
Let's name the function based on its functionality.

rata replied:
I think the name is the functionality: memberlist (ml) join loop is what this does

Is a very used pattern, like in Kubernetes code base. What name do you suggest?

rata replied:
@johananl friendly ping?

rata replied:
Done!

Renamed also `speakersIPsLoop()` to `mlUpdateSpeakers()`. Let me know if you think this name is good or some other name is better

johananl commented on internal/speakerlist/speakerlist.go:
```diff
@@ -0,0 +1,301 @@
+	// Update mlSpeakersIPs in the background.
+	go sl.speakersIPsLoop()
+
+	// Every one minute, try to rejoin.
```
Why loop based on a timer here? I suggest we look into the following alternative (if possible):

- Only try to rejoin when there are "unjoined" members.
- Attempts a rejoin ASAP following a failure event, possibly backing off exponentially.

rata replied:
Makes sense. We can make a node leave event trigger a re-join. But I don't think the second point helps.

For example, regarding the second point, N nodes leave, we try to join them all. Now only N-X are not joined and we do that until when? An exponential backoff? Then, after some time without those nodes joining, another node leaves, and you need to reset the backoff, right? It starts to feel complicated. Instead of exponential backoff, just retry every minute if there is something to do.

If you do it once a minute, only when there are nodes to join (i.e. it checks members and only tries to join when some IP is not member). 

I agree with doing things the way you propose, I just didn't find any easy/simple way to do that.

champtar replied:
@johananl what you want is more or less in #595, so let's clean this PR, merge it and have the patch release, then we can review 1 patch at a time #595

rata replied:
@johananl does what I said makes sense for you at least for now?

When we have a watch mechanism, I think, the only logic here will be to join the IPs that are not part of the cluster but returned by the watch mechanism. Maybe every one minute, maybe with some exponential backoff, we will need to see how the watch mechanism looks like.

Does it makes sense to leave it as it is for now? Or what would you change (without a watch)?

champtar commented on internal/speakerlist/speakerlist.go:
```diff
@@ -71,6 +77,16 @@ func New(logger gokitlog.Logger, nodeName, bindAddr, bindPort, secret, namespace
+	// K8s max node size is 5k, and this call back on node rejoin can be
+	// called 2-3 times (node change to Not Ready, then Ready again and
+	// sometimes another change is detected too), so we make it 5k * 3
+	sl.mlJoinCh = make(chan struct{}, 5000*3)
```
len should be 1, and to queue something:
```
select {
case sl.mlJoinCh <- struct{}{}:
    // some log
default:
    // some other log
}
```

rata replied:
>  This is worse than that, this is a split brain and you might have multiple nodes responding to ARP requests causing the tcp connection to switch nodes and get tcp reset

Yes, true. Didn't go into the details as I linked the issue. Tried to explain but the commit message was already big and explaining that in more detail was complicated while being clear, so I just explained that bit and let the link to the issue do the rest :)

champtar replied:
`sl.mlJoinCh = make(chan struct{}, 1)`
you understand me correctly ;)

johananl replied:
Fixed.

rata commented on internal/speakerlist/speakerlist.go:
```diff
@@ -31,9 +32,14 @@ type SpeakerList struct {
+	mlJoinCh  chan struct{}
+
+	mlMux         sync.Mutex // Mutex for mlSpeakersIPs
+	mlSpeakersIPs []string   // Speaker IPs
```
What do you mean with safer? I've stored []string because that is what we need for memberlist `Join()` function.

johananl replied:
By "safer" I meant using `net.IP` to ensure the strings are IP addresses. I've looked at `Join()` and I now see that it expects **hostnames**, therefore the use of strings.

rata replied:
Yes, and we are getting them via this function that return a slice of strings: https://github.com/metallb/metallb/blob/80bfe1cf45bf4338e6eced11ab7a88d3d2356663/internal/k8s/k8s.go#L268-L272.

So, IMHO, it seems safe. What do you think?

rata replied:
Ouch, just see it is marked as resolved :facepalm:

champtar commented on internal/speakerlist/speakerlist.go:
```diff
@@ -92,16 +108,155 @@ func (sl *SpeakerList) Start(client *k8s.Client) error {
+		return 0, nil
+	}
+
+	for i := 0; i < 2; i++ {
```
I'm not sure we should retry multiple time
To join the memberList you just need to join 1 of the members and you will rejoin the whole network.
Partial join means one node is really down I think

rata replied:
Down or not ready.

I'm fine without a retry, because we will retry in a minute anyways if something fails. Just that the time to join a node might be longer. Are you okay with that too? :)

champtar replied:
I prefer to have this fix simple (but correct) as we will likely remove most of it in the next minor release anyway

rata replied:
Cool, removed!

rata commented on internal/speakerlist/speakerlist.go:
```diff
@@ -0,0 +1,146 @@
+		mconfig.BindPort = mlport
+		mconfig.AdvertisePort = mlport
+	}
+	loggerout := gokitlog.NewStdlibAdapter(gokitlog.With(sl.l, "component", "MemberList"))
```
This is moving current code, see https://github.com/metallb/metallb/pull/662/files#diff-60053ad6fecb5a3cfabb6f3d9e720899L122

The second logger is used so memberlist uses the logger, instead of just writing to stdout. Not sure if it can be done easier. However, this is already on master, is the first commit that just moves the code

johananl replied:
Got it. I'll take a look to see if this makes sense. I understand if you don't want to touch this in the context of this PR though.

johananl replied:
Context: https://github.com/metallb/metallb/commit/734ee6743ebe9a8961a95a73ecb24b7f58f0e684

I'll leave this be for now and open an issue to look into the way we do logging for memberlist in more detail.

johananl replied:
https://github.com/metallb/metallb/issues/664

champtar commented on internal/speakerlist/speakerlist.go:
```diff
@@ -92,16 +108,155 @@ func (sl *SpeakerList) Start(client *k8s.Client) error {
+	}
+
+	sl.l.Log("op", "Member detection", "msg", "Sending message to re-process cluster members")
+	sl.mlJoinCh <- struct{}{}
```
see comment on channel size

rata replied:
Thanks!

rata commented on internal/speakerlist/speakerlist.go:
```diff
@@ -0,0 +1,146 @@
+	if bindPort != "" {
+		mlport, err := strconv.Atoi(bindPort)
+		if err != nil {
+			sl.l.Log("op", "startup", "error", "unable to parse ml-bindport", "msg", err)
```
Not sure, tbh. Probably a c&p mistake. Will check. Thanks for noticing! :)

champtar replied:
yup, look like a c&p mistake

johananl replied:
Fixed.

rata commented on internal/speakerlist/speakerlist.go:
```diff
@@ -71,6 +77,16 @@ func New(logger gokitlog.Logger, nodeName, bindAddr, bindPort, secret, namespace
+	// K8s max node size is 5k, and this call back on node rejoin can be
+	// called 2-3 times (node change to Not Ready, then Ready again and
+	// sometimes another change is detected too), so we make it 5k * 3
+	sl.mlJoinCh = make(chan struct{}, 5000*3)
```
You block on the main thread, so if a service type load balancer is created it won't receive an IP until this finishes and the main thread continues processing

I haven't meassured the impact on mem, I doubt it has any. Linux doesn't really allocate pages until you use them. Any suggestions on what I should check and how?

rata replied:
With champtar suggestion, we will never block. So, I'd mark this as resolved when doing that change. Consider it solved in the next push :-D

johananl replied:
@rata I've looked a bit deeper and realized a speaker pod only watches for its **own** `Node` object. Changes to other node objects aren't triggering `SetNode()`.

champtar commented:
@johananl that is why I wanted to review each commit separately, I can't address any of your reviews on my commit
@rata I'll let you fix my commit

rata commented:
@champtar 
>  This is worse than that, this is a split brain and you might have multiple nodes responding to ARP requests causing the tcp connection to switch nodes and get tcp reset

Yes, true. Didn't go into the details as I linked the issue. Tried to explain but the commit message was already big and explaining that in more detail was complicated while being clear, so I just explained that bit and let the link to the issue do the rest :)

champtar commented:
> @champtar
> 
> > This is worse than that, this is a split brain and you might have multiple nodes responding to ARP requests causing the tcp connection to switch nodes and get tcp reset
> 
> Yes, true. Didn't go into the details as I linked the issue. Tried to explain but the commit message was already big and explaining that in more detail was complicated while being clear, so I just explained that bit and let the link to the issue do the rest :)

No need to go into detail, but I read your message as 1 node is unusable, when we should read the full service might be unusable, kind of day and night

liuyuan10 commented on internal/speakerlist/speakerlist.go:
```diff
@@ -0,0 +1,301 @@
+func (sl *SpeakerList) mlSpeakers() ([]string, error) {
+	// This call blocks until we get a response of the API server.
+	// In the client-go version we are using, there is no way to use a
+	// context or timeout.
```
can we change Config.Timeout 
https://godoc.org/k8s.io/client-go/rest#Config

rata replied:
I think I checked this and it was removed in out client-go version, but I might have check it incorrectly (was kind of in a hurry :)). Will check again, thanks! :)

rata replied:
Oh, I think I remember now. The clientset used here is shared across the code and stored in a struct. I think that was the reason to not modify that.

Thanks!

rata commented on internal/speakerlist/speakerlist.go:
```diff
@@ -0,0 +1,301 @@
+			return
+		case <-ticker.C:
+			sl.mlJoin()
+		case <-sl.mlJoinCh:
```
Yeap, the list is stale. But if only one IP is from a speaker pod, it will join the memberlist cluster successfully and the list will be updated in the background. 

That would be improved, of course, if we rely on some watching mechanism.

rata replied:
Marked as resolved, as it is a known trade-off. Will be improved in future minor releases probably using a watch mechanism.

rata commented on internal/speakerlist/speakerlist.go:
```diff
@@ -0,0 +1,301 @@
+func New(logger gokitlog.Logger, nodeName, bindAddr, bindPort, secret, namespace, labels string, stopCh chan struct{}) (*SpeakerList, error) {
+	sl := SpeakerList{
+		l:         logger,
+		stopCh:    stopCh,
```
I agree. Not sure if it should be separate or not, though, as code wasn't very readable on my quick try, but maybe I was trying too quick. I left that for a following PR, though.

When I opened this PR [I created an issue][link] about leaving orphan goroutines. Currently the code on master is not doing anything about it. I think we should also create a wait group and actually wait for them.

[link]: https://github.com/metallb/metallb/issues/663

rata replied:
I added the issue to the 0.9.4 milestone.

rata replied:
@johananl as I mentioned in the comments, this PR **is** adding new go routines that are leaked. Cleaning them up is the same effort as fixing other leaking go routines in the speakerlist code.

I don't like adding new leaks either, but I created the issue for that reason, as I think this will be better fixed when this code is improved. Most go routines might not exist then, as we might use a simpler watch mechanism.

IMHO, at this point in time, I'd really like to release a patch release soon and work on further improvements for the next minor. IOW, I prefer to fix the leaking go routines for the 0.10 release.

WDYT?

rata replied:
@johananl ping?

johananl replied:
I'd like to get this released ASAP, too, but I also feel uncomfortable about consciously introducing leaks. Should I take a quick look at this and try to solve the problem?

johananl replied:
The following seems to work:

```diff
diff --git a/internal/speakerlist/speakerlist.go b/internal/speakerlist/speakerlist.go
index bd8c95c1..7cd31443 100644
--- a/internal/speakerlist/speakerlist.go
+++ b/internal/speakerlist/speakerlist.go
@@ -27,9 +27,16 @@ import (
 
 // SpeakerList represents a list of healthy speakers.
 type SpeakerList struct {
-	l         gokitlog.Logger
-	client    *k8s.Client
-	stopCh    chan struct{}
+	l      gokitlog.Logger
+	client *k8s.Client
+
+	// A channel callers can use to stop the SpeakerList.
+	stopCh chan struct{}
+	// A channel SpeakerList uses to stop child goroutines.
+	childrenStopCh chan struct{}
+	// A WaitGroup for tracking child goroutines.
+	childrenWG sync.WaitGroup
+
 	namespace string
 	labels    string
 
@@ -45,10 +52,11 @@ type SpeakerList struct {
 // New creates a new SpeakerList and returns a pointer to it.
 func New(logger gokitlog.Logger, nodeName, bindAddr, bindPort, secret, namespace, labels string, stopCh chan struct{}) (*SpeakerList, error) {
 	sl := SpeakerList{
-		l:         logger,
-		stopCh:    stopCh,
-		namespace: namespace,
-		labels:    labels,
+		l:              logger,
+		stopCh:         stopCh,
+		childrenStopCh: make(chan struct{}),
+		namespace:      namespace,
+		labels:         labels,
 	}
 
 	if namespace == "" || labels == "" || bindAddr == "" {
@@ -143,24 +151,27 @@ func (sl *SpeakerList) Start(client *k8s.Client) {
 	sl.mlSpeakerIPs = iplist
 	sl.mlMux.Unlock()
 
-	// Update mlSpeakerIPs in the background.
-	go sl.updateSpeakerIPs()
+	sl.childrenWG.Add(3)
 
-	go sl.memberlistWatchEvents()
-	go sl.joinMembers()
+	// Update mlSpeakerIPs in the background.
+	go sl.updateSpeakerIPs(sl.childrenStopCh)
+	go sl.memberlistWatchEvents(sl.childrenStopCh)
+	go sl.joinMembers(sl.childrenStopCh)
 }
 
 // updateSpeakerIPs runs forever updating the sl.mlSpeakerIPs slice with the
 // IPs of all the speaker pods. As this queries the API server, it waits at
 // least 5 minutes between queries, to avoid self-induced API server DoS (should
 // be trated with care).
-func (sl *SpeakerList) updateSpeakerIPs() {
+func (sl *SpeakerList) updateSpeakerIPs(stop <-chan struct{}) {
 	for {
 		// This either stops the loop or makes sure to wait at least 5
 		// minutes before making another call to sl.mlSpeakers(), which
 		// queries the API server.
 		select {
-		case <-sl.stopCh:
+		case <-stop:
+			sl.l.Log("op", "shutdown", "msg", "stopping updateSpeakerIPs")
+			sl.childrenWG.Done()
 			return
 		case <-time.After(5 * time.Minute):
 			// This blocks until the API server responds.
@@ -198,7 +209,7 @@ func (sl *SpeakerList) mlSpeakers() ([]string, error) {
 	return iplist, nil
 }
 
-func (sl *SpeakerList) joinMembers() {
+func (sl *SpeakerList) joinMembers(stop <-chan struct{}) {
 	// Every one minute, try to rejoin.
 	// This will join nodes that leave the cluster only for a few seconds every 1
 	// minute, while discovering new IPs (updated by sl.speakersIPsLoop()) might
@@ -207,7 +218,9 @@ func (sl *SpeakerList) joinMembers() {
 
 	for {
 		select {
-		case <-sl.stopCh:
+		case <-stop:
+			sl.l.Log("op", "shutdown", "msg", "stopping joinMembers")
+			sl.childrenWG.Done()
 			return
 		case <-ticker.C:
 			sl.mlJoin()
@@ -295,6 +308,11 @@ func (sl *SpeakerList) Stop() {
 		return
 	}
 
+	sl.l.Log("op", "shutdown", "msg", "stopping child goroutines")
+	close(sl.childrenStopCh)
+	sl.childrenWG.Wait()
+	sl.l.Log("op", "shutdown", "msg", "child goroutines stopped")
+
 	sl.l.Log("op", "shutdown", "msg", "leaving memberlist cluster")
 	err := sl.ml.Leave(time.Second)
 	sl.l.Log("op", "shutdown", "msg", "left memberlist cluster", "error", err)
@@ -306,13 +324,15 @@ func event2String(e memberlist.NodeEventType) string {
 	return []string{"NodeJoin", "NodeLeave", "NodeUpdate"}[e]
 }
 
-func (sl *SpeakerList) memberlistWatchEvents() {
+func (sl *SpeakerList) memberlistWatchEvents(stop <-chan struct{}) {
 	for {
 		select {
 		case e := <-sl.mlEventCh:
 			sl.l.Log("msg", "node event - forcing sync", "node addr", e.Node.Addr, "node name", e.Node.Name, "node event", event2String(e.Event))
 			sl.client.ForceSync()
-		case <-sl.stopCh:
+		case <-stop:
+			sl.l.Log("op", "shutdown", "msg", "stopping memberlistWatchEvents")
+			sl.childrenWG.Done()
 			return
 		}
 	}
```

We create a dedicated "children stop channel" for the SpeakerList's "background processes". We pass this channel to all child goroutines as an argument. When we receive a signal on the "main" stop channel, we close the children's channel and wait for the wait group.

We can probably tweak the logs a bit though.

johananl replied:
I've pushed the code here: https://github.com/kinvolk/metallb/commit/a6d84149ba6c13076face44af1acc75c20872b78

rata commented on internal/speakerlist/speakerlist.go:
```diff
@@ -0,0 +1,301 @@
+	// This will join nodes that leave the cluster only for a few seconds every 1
+	// minute, while discovering new IPs (updated by sl.speakersIPsLoop()) might
+	// take a while longer.
+	ticker := time.NewTicker(1 * time.Minute)
```
Oh, maybe that is simpler. Good point. Let's see if we want to watch for k8s events or not, and if we are going this way I'll check if that makes the code simpler. Thanks! :)

champtar replied:
If we watch for k8s objects it's going to be closer and closer to my PR or am I missing something ?

liuyuan10 replied:
I think rata@ doesn't want to introduce extra resource objects in manifests (i.e. the extra service in your PR), which is hard to remove if we ever want to remove it. That's the key difference I see it.

champtar replied:
It's not a problem if we don't remove it, and I'm providing a clean way to remove it ...
Right now we only have pod list permission, so we can't change that for a patch release.

rata replied:
Yes, I don't want to include a service if we will remove it in the next release. Watch permissions, if and only if it is more clear that we will keep them, seems better, IMHO.

champtar replied:
Still, can't change the manifests in a patch release IMO.

I still believe that we should use the best tool for the job (here watching service), but if that's the only blocking point, I can have the k8s client go watching pods in SpeakerList

rata replied:
Switched to `time.After()`, it indeed makes it way simpler. Thanks!

champtar replied:
Advantage of ticker is that it run every X exactly, so it sometimes make things easier to investigate weird behavior (I'm on my phone so can't see code before / after)

rata replied:
Not really. If the ticker is every X and when it ticks you call a function that takes 2X time, then you will call the function the first time and receive the second time after a few milliseconds.

The idea here is to avoid doing that, just make sure to not call more than every 5 minutes. 

If the API server is down, for example, your calls can take long and you might end up calling the API server more than expected. I think using `time.After()` is quite simpler here.

rata commented:
@liuyuan10 thanks a lot for the review! :)

rata commented on internal/speakerlist/speakerlist.go:
```diff
@@ -0,0 +1,146 @@
+		sha := sha256.New()
+		mconfig.SecretKey = sha.Sum([]byte(secret))[:16]
+	}
+	// ChannelEventDelegate hint that it should not block, so make mlEventCh 'big'
```
From https://pkg.go.dev/github.com/hashicorp/memberlist@v0.1.7?tab=doc#ChannelEventDelegate:

> Care must be taken that events are processed in a timely manner from the channel, since this delegate will block until an event can be sent. 

I'm not 100% sure what this would cause, maybe the node might be considered down. But:
 * This is the code we already have in master, we are just moving it in the commit
 * It is probably not a big issue, because events are processed quite fast: https://github.com/metallb/metallb/pull/662/files#diff-b7338554f9f6fe870766bf340b113a65R291-R300

I can create an issue to see the pros/cons of moving to https://pkg.go.dev/github.com/hashicorp/memberlist@v0.1.7?tab=doc#EventDelegate

Maybe @champtar is already aware of the pros/cons?

champtar replied:
From the top of my head, I think I've rework how I use those events to make sure I don't block, but it was better (or easier) to use the channel that the function call.

johananl replied:
@rata I'm OK with leaving this for now since we're just moving this code around, however I'd rather not have bits in our codebase we don't understand, and in general this feels a bit hacky to me. I suggest leaving a TODO with a link to an issue so that we don't forget to handle this later.

rata replied:
Added a TODO comment pointing to this issue: https://github.com/metallb/metallb/issues/716

rata commented on internal/speakerlist/speakerlist.go:
```diff
@@ -0,0 +1,146 @@
+	if err != nil {
+		sl.l.Log("op", "startup", "error", err, "msg", "failed to create memberlist")
+		return nil, err
+	}
```
Fixed, thanks!

rata commented on internal/speakerlist/speakerlist.go:
```diff
@@ -0,0 +1,146 @@
+	return &sl, nil
+}
+
+// Start starts the needed goroutines
```
Improved the comment now, thanks. Let me know if it makes sense now :)

rata commented on internal/speakerlist/speakerlist.go:
```diff
@@ -0,0 +1,146 @@
+	}
+
+	mconfig := memberlist.DefaultLANConfig()
+	// mconfig.Name MUST be spec.nodeName, as we will match it against Enpoints nodeName in usableNodes()
```
I think I fixed them all (will be visible in the next push). Thanks!

rata commented on internal/speakerlist/speakerlist.go:
```diff
@@ -0,0 +1,146 @@
+	}
+
+	mconfig := memberlist.DefaultLANConfig()
+	// mconfig.Name MUST be spec.nodeName, as we will match it against Enpoints nodeName in usableNodes()
```
Added locally, thanks!

rata commented on internal/speakerlist/speakerlist.go:
```diff
@@ -0,0 +1,146 @@
+	"github.com/hashicorp/memberlist"
+)
+
+// SpeakerList gives you the list of healthy speaker
```
Added locally. Thanks!

rata commented on internal/speakerlist/speakerlist.go:
```diff
@@ -0,0 +1,146 @@
+	stopCh    chan struct{}
+	namespace string
+	labels    string
+	// following fields are nil when MemberList is disabled
```
Fixed locally, thanks!

rata commented on internal/speakerlist/speakerlist.go:
```diff
@@ -0,0 +1,146 @@
+
+// Start starts the needed goroutines
+func (sl *SpeakerList) Start(client *k8s.Client) error {
+	sl.client = client
```
Probably. I'm trying to see how to fix this, but doesn't seem easy. I'm unsure if we should use the patch that moves it to another package, then :-/

The issue is the following:
* The `newController()` function [takes the sList as param](https://github.com/metallb/metallb/pull/662/files#diff-60053ad6fecb5a3cfabb6f3d9e720899R109)
 * The `newController()` function uses it [to create the layer 2 controller](https://github.com/metallb/metallb/pull/662/files#diff-60053ad6fecb5a3cfabb6f3d9e720899L232-L251)
 * The new Kubernetes client [uses the just created controller to set the callbacks](https://github.com/metallb/metallb/pull/662/files#diff-60053ad6fecb5a3cfabb6f3d9e720899L160-L162)

Then, we have a loop if we add the kubernetes client to  `speakerlist.Start()`:
 * To create a kubernetes client we need a controller
 * To create a controller we need a speaker list
 * To create a speakerlist we need a kubernetes client

I don't like it, but I think there is no easy way around this. So, I'll leave it. It is clear that we need to rework the layer 2 code afterwards for this and for other things.

What do you think?

champtar replied:
This cyclic dependency is removed in my other PR, please leave it like that.

rata replied:
Cool. I think the dependency can also be removed by also another improvement: moving memberlist inside layer 2 code, not leak a layer 2 implementation detail to the generic, protocol agnostic code.

champtar replied:
I don't know if it could be useful to BGP or not, Maybe it's better to implement BFD, but I might have to look into that in the coming months.

rata replied:
Added a big fat comment explaining this. Thanks!

rata commented on internal/speakerlist/speakerlist.go:
```diff
@@ -0,0 +1,301 @@
+}
+
+// Start starts the needed goroutines
+func (sl *SpeakerList) Start(client *k8s.Client) error {
```
Thanks, fixed!

rata commented on internal/speakerlist/speakerlist.go:
```diff
@@ -0,0 +1,146 @@
+	mList     *memberlist.Memberlist
+}
+
+// New creates a new SpeakerList
```
Fixed, thanks!

rata commented on internal/speakerlist/speakerlist.go:
```diff
@@ -0,0 +1,146 @@
+	stopCh    chan struct{}
+	namespace string
+	labels    string
+	// following fields are nil when MemberList is disabled
```
Fixed, thanks!

rata commented on internal/speakerlist/speakerlist.go:
```diff
@@ -0,0 +1,146 @@
+	if bindPort != "" {
+		mlport, err := strconv.Atoi(bindPort)
+		if err != nil {
+			sl.l.Log("op", "startup", "error", "unable to parse ml-bindport", "msg", err)
```
It is consistent now to start with lower case. Thanks!

rata commented:
@johananl @champtar PTAL.

Just fixed most of the comments. There are a few open that need more discussion :)

johananl reviewed (commented):
Thanks for the effort @rata.

champtar commented:
I'm fine with having code review, but any plan to ship something at some point ?
- we have bug reports that might be fixed by this PR
- a good part of this code is changed/going away in my other commits in #595

rata commented:
> I'm fine with having code review, but any plan to ship something at some point ?

Yes, of course :). I'd like to release in the next few weeks. I'll try to address @johananl review and merge when it is ready.

rata commented on internal/speakerlist/speakerlist.go:
```diff
@@ -93,24 +103,156 @@ func New(logger gokitlog.Logger, nodeName, bindAddr, bindPort, secret, namespace
+		case <-sl.stopCh:
+			return
+		case <-time.After(5 * time.Minute):
+			// This blocks until the apiserver responds.
```
Applied it locally, thanks!

rata commented on internal/speakerlist/speakerlist.go:
```diff
@@ -0,0 +1,155 @@
+
+	// The following fields are nil when memberlist is disabled.
+	mlEventCh chan memberlist.NodeEvent
+	mList     *memberlist.Memberlist
```
Makes sense. It is now just `ml` :)

rata commented on internal/speakerlist/speakerlist.go:
```diff
@@ -0,0 +1,297 @@
+	mlJoinCh  chan struct{}
+
+	mlMux         sync.Mutex // Mutex for mlSpeakersIPs.
+	mlSpeakersIPs []string   // Speaker pods IPs.
```
I think I never thought about this before, but makes lot of sense. Thanks! :)

Updated both :)

rata commented on internal/speakerlist/speakerlist.go:
```diff
@@ -93,24 +103,156 @@ func New(logger gokitlog.Logger, nodeName, bindAddr, bindPort, secret, namespace
+// IPs of all the speaker pods. As this queries the API server, it waits at
+// least 5 minutes between queries, to avoid self-induced API server DoS (should
+// be trated with care).
+func (sl *SpeakerList) speakersIPsLoop() {
```
Ouch, I'm just seeing this. I renamed before seeing this comment to `mlUpdateSpeakers()`. Let me know if I should rename to this

rata replied:
To minimize possible iterations, I've update to the name you suggested: updateSpeakerIPs()

rata commented on internal/speakerlist/speakerlist.go:
```diff
@@ -0,0 +1,297 @@
+	sl.l.Log("op", "shutdown", "msg", "memberlist shutdown", "error", err)
+}
+
+func event2String(e memberlist.NodeEventType) string {
```
just to be clear: this is what is already merged in main branch.

Converts `memberlist.NodeEventType` in a string to print the event as a string (the struct is not printed nicely otherwise, IIRC).

This function is used to print a log line (string) when a node joins or leaves.

rata replied:
@johananl re-ping?

rata commented on internal/speakerlist/speakerlist.go:
```diff
@@ -0,0 +1,297 @@
+	sl.mlMux.Unlock()
+
+	// Update mlSpeakersIPs in the background.
+	go sl.speakersIPsLoop()
```
I couldn't decide if doing it here or `Start()` and chose here as I though of the `speakersIPsLoop()` as an implementation detail of how we are discovering new IPs to join now.

But moved it to `Start()` :)

rata commented on internal/speakerlist/speakerlist.go:
```diff
@@ -0,0 +1,155 @@
+	namespace string
+	labels    string
+
+	// The following fields are nil when memberlist is disabled.
```
It is an _optional_ implementation detail. You can use it or not, you can choose in the manifest (see instructions here: https://metallb.universe.tf/release-notes/#version-0-9-2). If memberlist is disabled, those vars are nil.

I guess that was the confusion? Or what do you mean here with "interchangeably"?

johananl replied:
I understand both "memberlist" and "speakerlist". What I find confusing/inconsistent is that we seem to be referring to this *entire feature* (namely the "fast dead node detection feature") sometimes as "memberlist" and other times as "speakerlist".

For example, let's look at this:

https://github.com/kinvolk/metallb/blob/38bd40820b57c11fb8a38de39b456add61fc3075/internal/speakerlist/speakerlist.go#L28-L43

Here we have a type called `speakerlist.SpeakerList` which uses **memberlist** in its fields. This fact suggests that memberlist is an implementation detail of speakerlist, i.e. we have a feature called "speakerlist" which uses memberlist under the hood.

Now let's look at this:

https://github.com/kinvolk/metallb/blob/38bd40820b57c11fb8a38de39b456add61fc3075/speaker/main.go#L72-L76

The user-facing knobs mention **memberlist** not SpeakerList, which I find inconsistent with the above.

To try and boil this down to a one-liner, I'm trying to understand if we should call the fast dead node detection feature "memberlist" or "speakerlist".

rata replied:
Would it make any difference for this PR?

AFAIK, we shouldn't change the user-visible knobs for a patch release (unless is _really_ important). And if the documentation should mention to this feature in one one way or the other, what difference would it make to this PR?

It seems a legit concern, but I don't see why it would make sense to block this PR until then.

rata replied:
Okay. Will mark this as solved, thanks!

rata commented on internal/speakerlist/speakerlist.go:
```diff
@@ -0,0 +1,297 @@
+	// be called from mlJoin() with a context (timeout) and if doesn't
+	// answer, use sl.mlSpeakersIPs.  If it doesn't time out, just update
+	// sl.mlSpeakersIPs with the answer and continue to do the join.
+	iplist, err := sl.client.GetPodsIPs(sl.namespace, sl.labels)
```
Ok. I added a new commit to do that

rata commented on speaker/layer2_controller.go:
```diff
@@ -39,23 +38,15 @@ func (c *layer2Controller) SetConfig(log.Logger, *config.Config) error {
-		}
-	}
-
+func usableNodes(eps *v1.Endpoints, usableSpeakers map[string]bool) []string {
```
Renamed, while looking at the code fixed some bugs created when moving it to speakerlist package.

rata commented on internal/speakerlist/speakerlist.go:
```diff
@@ -0,0 +1,155 @@
+}
+
+// New creates a new SpeakerList and returns a pointer to it.
+func New(logger gokitlog.Logger, nodeName, bindAddr, bindPort, secret, namespace, labels string, stopCh chan struct{}) (*SpeakerList, error) {
```
I don't think I agree with this. At least with how we have things now.

The channel is created out of here and is passed to `internal/k8s` too. Just closing that channel signals all (speakerlist, intarnal/k8s, etc.) to close. Please note that the close method on a channel is sent to everyone receiving on that channel.

If the new is going to create a channel here instead, then we would need to keep track of several channels and close them all in the calling places. That seems quite impractical, IMHO, and not a minor change.

We can revisit this when doing a new minor release, IMHO (please create an issue if you think it is important to not forget about this. IMHO, if this is important it will be obvious when improving the code).

Marking as resolved for those reasons, feel free to re-open if I missed something :)

johananl replied:
All good, I guess I looked over the fact that we're using this channel with the fan-out pattern.

rata commented:
@johananl thanks for the review. PTAL

I think I addressed all the comments. Thanks a lot for the suggestions, I applied them all and amended them to the proper commit.

johananl commented on internal/speakerlist/speakerlist.go:
```diff
@@ -0,0 +1,319 @@
+}
+
+func (sl *SpeakerList) joinMembers() {
+	// Every one minute, try to rejoin.
+	// This will join nodes that leave the cluster only for a few seconds every 1
+	// minute, while discovering new IPs (updated by sl.speakersIPsLoop()) might
+	// take a while longer
```
```suggestion
	// Every one minute, try to rejoin.
	// This joins nodes that leave the cluster for just a few seconds.
	// Discovering new IPs (updated by sl.updateSpeakerIPs()) might take a while
	// longer.
```

- "every 1 minute" is repeated twice.
- Change future tense to present.
- "For just a few seconds" is more idiomatic English IMO.
- Remove ambiguous "while".
- Update `speakersIPsLoop` to `sl.updateSpeakerIPs` which is the current name of the method.

In addition to the above, what does the word "longer" relate to? Discovering new IPs takes longer than *what*?

johananl commented on internal/speakerlist/speakerlist.go:
```diff
@@ -0,0 +1,319 @@
+	return members
+}
+
+// mlJoin joins this memberlist cluster with other speaker pods that are not
+// currently members of this cluster. In other words, it does a memberlist.Join()
+// with the IPs in mlSpeakerIPs that are not members of the cluster.
```
```suggestion
// mlJoin joins speaker pods that are not members of this cluster
// to the cluster. It performs a memberlist.Join() with the IPs in
// mlSpeakerIPs that are not members of the cluster.
```

- "Join a cluster with a pod" is IMO confusing both at the English level and the technical level.
- "In other words" doesn't contribute to the meaning of this docstring IMO.
- I find "perform" better in the context of a *process* ("I perform a process" vs. "I do a process").

johananl commented on internal/speakerlist/speakerlist.go:
```diff
@@ -0,0 +1,319 @@
+	defer sl.mlMux.Unlock()
+
+	for _, ip := range sl.mlSpeakerIPs {
+		// If an IP is not member of the cluster, add it to joinIPs.
```
```suggestion
		// If an IP is not a member of the cluster, add it to joinIPs.
```

johananl commented on speaker/layer2_controller.go:
```diff
@@ -39,23 +38,19 @@ func (c *layer2Controller) SetConfig(log.Logger, *config.Config) error {
-
+// The speakers parameter is a map with the node name as key and the readiness
+// status as value (true means ready, false means not ready).
+// If the speakers map is nil, it is completely ignored to return the
+// usableNodes.
```
```suggestion
// If the speakers map is nil, it is ignored.
```

johananl commented on speaker/main.go:
```diff
@@ -429,3 +364,9 @@ type Protocol interface {
 }
+
+// Speakerlist represents a list of healthy speakers.
+type Speakerlist interface {
```
Is there a reason we use `Speakerlist` here but `SpeakerList` elsewhere?

rata replied:
No, switched them all to SpeakerList

johananl commented on internal/speakerlist/speakerlist.go:
```diff
@@ -0,0 +1,319 @@
+	// while waiting for the join operation to complete.
+	sl.mlJoinCh = make(chan struct{}, 1)
+
+	// ChannelEventDelegate hint that it should not block, so make mlEventCh
```
```suggestion
	// ChannelEventDelegate hints that it should not block, so make mlEventCh
```

rata commented on internal/speakerlist/speakerlist.go:
```diff
@@ -0,0 +1,319 @@
+	return &sl, nil
+}
+
+// Start initializes the speakerlist. This functions must be called before using
+// other SpeakerList methods.
```
Switched to that locally, will be visible on the next push. Thanks!

rata commented on internal/speakerlist/speakerlist.go:
```diff
@@ -0,0 +1,319 @@
+	// TODO: the k8s client parameter ideally, should be a parameter to
+	// New(). However, that is not possible today because:
+	//
+	// 1. The newController() function takes the sList as param[1]
```
Fixed, thanks!

rata commented on internal/speakerlist/speakerlist.go:
```diff
@@ -0,0 +1,319 @@
+	// New(). However, that is not possible today because:
+	//
+	// 1. The newController() function takes the sList as param[1]
+	// 2. The newController() function uses it to create the layer 2 controller [2]
```
Replaced, `it` meant SpeakerList

rata commented:
@johananl PTAL. I think I address your last remaining comments.

FYI, I amended your github suggestions to the respective commits that introduced them. That is why you won't see a commit doing them, but they are addressed :)

rata commented on internal/speakerlist/speakerlist.go:
```diff
@@ -0,0 +1,319 @@
+	// 2. The newController() function uses it to create the layer 2 controller [2]
+	// 3. The new Kubernetes client uses the just created controller to set the callbacks [3]
+	//
+	// Then, we have a loop if we add the kubernetes client to speakerlist.New():
```
Oops, fixed. Thanks!

johananl commented:
>FYI, I amended your github suggestions to the respective commits that introduced them. That is why you won't see a commit doing them, but they are addressed :)

That's fine, I never actually use the "commit" UI option. I use the suggestions feature just for the diff display.

johananl commented on internal/speakerlist/speakerlist.go:
```diff
@@ -0,0 +1,319 @@
+	//
+	// 1. The newController() function takes the sList as param[1]
+	// 2. The newController() function uses it to create the layer 2 controller [2]
+	// 3. The new Kubernetes client uses the just created controller to set the callbacks [3]
```
I would say "newly-created controller". "the just-created controller" sounds like weird English to me. Anyway, let's not spend any more time on this.

johananl reviewed (approved):
LGTM. Let's :ship: it.
